src/main/java/org/opensearch/ad/cluster/ADDataMigrator.java

```diff
@@ -0,0 +1,288 @@
+    private final AtomicBoolean dataMigrated;
+
+    public ADDataMigrator(
+        Client client,
+        ClusterService clusterService,
+        NamedXContentRegistry xContentRegistry,
+        AnomalyDetectionIndices detectionIndices
+    ) {
+        this.client = client;
+        this.clusterService = clusterService;
+        this.xContentRegistry = xContentRegistry;
+        this.detectionIndices = detectionIndices;
+        this.dataMigrated = new AtomicBoolean(false);
+    }
+
+    /**
+     * Migrate AD data. Currently only need to migrate detector internal state {@link DetectorInternalState}
+     */
+    public void migrateData() {
+        if (!dataMigrated.getAndSet(true)) {
+            logger.info("Start migrating AD data");
+
+            if (!detectionIndices.doesAnomalyDetectorJobIndexExist()) {
+                logger.info("AD job index doesn't exist, no need to migrate");
+                return;
+            }
+
+            if (detectionIndices.doesDetectorStateIndexExist()) {
+                migrateDetectorInternalStateToRealtimeTask();
+            } else {
+                // If detection index doesn't exist, create index and backfill realtime task.
+                detectionIndices.initDetectionStateIndex(ActionListener.wrap(r -> {
+                    if (r.isAcknowledged()) {
+                        logger.info("Created {} with mappings.", CommonName.DETECTION_STATE_INDEX);
+                        migrateDetectorInternalStateToRealtimeTask();
+                    } else {
+                        String error = "Create index " + CommonName.DETECTION_STATE_INDEX + " with mappings not acknowledged";
+                        logger.warn(error);
+                    }
+                }, e -> {
+                    if (ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException) {
+                        migrateDetectorInternalStateToRealtimeTask();
+                    } else {
+                        logger.error("Failed to init anomaly detection state index", e);
+                    }
+                }));
+            }
+        }
+    }
+
+    /**
+     * Migrate detector internal state to realtime task.
+     */
+    public void migrateDetectorInternalStateToRealtimeTask() {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+            .query(new MatchAllQueryBuilder())
+            .size(MAX_DETECTOR_UPPER_LIMIT);
+        SearchRequest searchRequest = new SearchRequest(ANOMALY_DETECTOR_JOB_INDEX).source(searchSourceBuilder);
+        client.search(searchRequest, ActionListener.wrap(r -> {
+            if (r == null || r.getHits().getTotalHits() == null || r.getHits().getTotalHits().value == 0) {
+                logger.info("No anomaly detector job found, no need to migrate");
+                return;
+            }
+            ConcurrentLinkedQueue<AnomalyDetectorJob> detectorJobs = new ConcurrentLinkedQueue<>();
```

Comment:
Can you explain how/why there would be multiple jobs for a detector? And if so, why migrate all of them, instead of just the latest?

Reply:
That's not for one detector. We need to migrate all AD jobs for all old detectors. One detector can only have only one AD job as the job doc id is detector id.

Reply:
Got it. I think I confused myself when testing it out and seeing the `AD data migration done` for each detector, thinking it was running this method per detector. I think this was some earlier version of the changes and I can't reproduce now. Can ignore.